src/Common/src/TypeSystem/Interop/IL/Marshaller.cs

```diff
@@ -778,23 +987,74 @@ protected virtual void EmitReverseArgumentMarshallingIL()
                 EmitMarshalArgumentNativeToManaged();
             }
         }
+
+        protected virtual void EmitForwardElementMarshallingIL()
+        {
+            if (In)
+                EmitMarshalElementManagedToNative();
+            else
+                EmitMarshalElementNativeToManaged();
+        }
+
+        protected virtual void EmitReverseElementMarshallingIL()
+        {
+            if (In)
+                EmitMarshalElementNativeToManaged();
+            else
+                EmitMarshalElementManagedToNative();
+        }
+
         protected virtual void EmitMarshalReturnValueManagedToNative()
         {
+            ILEmitter emitter = _ilCodeStreams.Emitter;
+            _vManaged = emitter.NewLocal(ManagedParameterType);
+            _vNative = emitter.NewLocal(NativeParameterType);
+            _ilCodeStreams.ReturnValueMarshallingCodeStream.EmitStLoc(_vNative);
+
+            EmitSpaceAndContentsNativeToManaged();
+
+            _ilCodeStreams.ReturnValueMarshallingCodeStream.EmitLdLoc(_vManaged);
+
+        }
+
+        protected virtual void EmitLoadArguments()
+        {
+            ILEmitter emitter = _ilCodeStreams.Emitter;
+            ILCodeStream marshallingCodeStream = _ilCodeStreams.MarshallingCodeStream;
+
+            _vManaged = emitter.NewLocal(ManagedParameterType);
+            if (IsByRef)
+            {
+                _vNative = emitter.NewLocal(((PointerType)NativeParameterType).ParameterType);
+            }
+            else
+            {
+                _vNative = emitter.NewLocal(NativeParameterType);
+            }
+            marshallingCodeStream.EmitLdArg(PInvokeParameterMetadata.Index - 1);
+            marshallingCodeStream.EmitStLoc(_vManaged);
         }
+
         protected virtual void EmitMarshalArgumentManagedToNative()
         {
+            EmitLoadArguments();
             //
             // marshal
             //
-            if (In)
+            if (IsByRef && !In)
             {
-                EmitSpaceAndContentsManagedToNative();
+                EmitReInitNativeContents();
             }
             else
             {
-                EmitSpaceManagedToNative();
+                EmitSpaceAndContentsManagedToNative();
```

Comment:
I don't like the Space and Contents name too :) But named it to have parity with .NET Native and CoreCLR. How about AllocAndTransformManagedToNative  since Marshal is used in other context?